ncbi/src/main/java/org/reactome/release/NCBIEntry.java

```diff
@@ -0,0 +1,201 @@
+	 * @return Set of Reactome Events which are the top level pathways
+	 */
+	public Set<ReactomeEvent> getTopLevelPathways(Session graphDBSession) {
+		return uniProtReactomeEntry.getTopLevelPathways(graphDBSession);
+	}
+
+	/**
+	 * Retrieves the list of NCBI Entry objects from UniProt entries in the Reactome graph database
+	 * which have NCBI Gene identifiers
+	 * @param graphDBSession Neo4J Driver Session object for querying the graph database
+	 * @return List of NCBI Entry objects
+	 */
+	public static List<NCBIEntry> getUniProtToNCBIGeneEntries(Session graphDBSession) {
+		logger.info("Generating UniProt accession to NCBI Gene mapping");
+
+		StatementResult result = graphDBSession.run(
+			String.join(System.lineSeparator(),
+				"MATCH (rgp:ReferenceGeneProduct)-[:referenceDatabase]->(rd:ReferenceDatabase)",
+				"MATCH (rgp)-[:referenceGene]->(rds:ReferenceDNASequence)",
+				"WHERE rd.displayName = 'UniProt' AND rds.databaseName = 'NCBI Gene'",
+				"RETURN rgp.dbId, rgp.displayName, rgp.identifier, rds.identifier",
+				"ORDER BY rgp.identifier;"
+			)
+		);
+
+		Map<UniProtReactomeEntry, Set<String>> uniprotToNCBIGene = new HashMap<>();
+		while (result.hasNext()) {
+			Record record = result.next();
+			long uniprotDbId = record.get("rgp.dbId").asLong();
+			String uniprotDisplayName = record.get("rgp.displayName").asString();
+			String uniprotAccession = record.get("rgp.identifier").asString();
+			String ncbiGeneID = record.get("rds.identifier").asString();
+
+			UniProtReactomeEntry uniprot = UniProtReactomeEntry.get(uniprotDbId, uniprotAccession, uniprotDisplayName);
+			Set<String> ncbiGeneIDs = uniprotToNCBIGene.computeIfAbsent(uniprot, k -> new HashSet<>());
+			ncbiGeneIDs.add(ncbiGeneID);
+		}
+
+		List<NCBIEntry> ncbiEntries = new ArrayList<>();
+		for (UniProtReactomeEntry uniprot : uniprotToNCBIGene.keySet()) {
+			ncbiEntries.add(
+				new NCBIEntry(uniprot, uniprotToNCBIGene.get(uniprot))
+			);
+		}
+		Collections.sort(ncbiEntries);
+
+		logger.info("Finished generating UniProt accession to NCBI Gene mapping");
+
+		return ncbiEntries;
+	}
+
+	/**
+	 * Compares UniProt accession values of this object and parameter
+	 * @param o NCBIEntry object to compare
+	 * @return Value of String compare between this UniProt accession and the parameter's UniProt accession
+	 */
+	@Override
+	public int compareTo(@Nonnull NCBIEntry o) {
+		return this.getUniprotAccession().compareTo(o.getUniprotAccession());
+	}
+
+	/**
+	 * Checks equality based on object type and value of UniProt accession, UniProt display name, and NCBI Gene ids
+	 * @param o Object to check for equality with the calling NCBI Entry object.
+	 * @return <code>true</code> if the same object or an NCBI Entry object with the same UniProt accession,
+	 * UniProt display name, and NCBI Gene ids.  Returns <code>false</code> otherwise.
+	 */
+	@Override
+	public boolean equals(Object o) {
```

Comment:
similar as above: variable naming isn't descriptive. As well, this code seems similar to the one above (though I haven't checked if it is or not). Would it make sense to create a 'utilities' class for this step for re-used code?

Reply:
A Utilities class _could_ be created that has equality methods comparing instances of various objects, but that's not as object-oriented as having the objects _themselves_ know how to check for equality. ;)
Also, there are libraries that do object comparison and rely on you implementing your own `equals` method for your classes. Failure to implement `equals` means object comparison might not work the way you want/expect it to. I've noticed that Joel used a `TreeMap` in a few places. Since that structure is implemented using a Red-Black tree, it is necessary to be able to compare objects properly.

Reply:
Yeah, unfortunately, there is some duplication in the patterns for the boilerplate of the `equals` methods for the different objects, but it is kind of necessary to ensure objects of those classes knows how to be compared.